File: matlab/function_EMCCD_CMOS_calib.py
import os
import warning
import numpy as np
from scipy.stats import  linregress

def EMCCD_CMOS_calib(folder = './',csv = True, ROIsize=9):

    #Extracting images and csv files
    files_tif = []
    files_csv = []
    # ~ filenames = []
    for file in os.listdir(folder):
        if file.endswith('.tif'):
            files_tif.append(file)
            # ~ filenames.append(file.split('_')[0])
        elif file.endswith('.csv'):
            files_csv.append(file)
            
    
    # ~ filenames = list(set(filenames))
    # ~ filenames.sort()
    files_tif.sort()
    files_csv.sort()
    #files = {name: [] for name in filenames}
        
    # Let us subdivide files with name of the cell
    # ~ for file in files_tif:
        # ~ filename = file.split('_')[0]
        # ~ files[filename].append(file)
    
    files = files_tif
    filesCMOS = [file for file in files if file.find('CMOS')>=0]
    if len(filesCMOS) == 0:
        raise 'Sorry, there are no files with CMOS in their name.'
    
    
    filesEMCCD = [file for file in files if file.find('EMCCD')>=0]
    if len(filesEMCCD)==0:
        raise 'We stop, no files with EMCCD in their name, for '+name+'.'
        
    
    ptsC = []
    ptsE = []
    for i,names in enumerate(zip(filesCMOS,filesEMCCD)): #real traces
        fCMOS, fEMCCD = names
    
        # Change these lines to appropiate reading of tif files:
        CMOS = imread(folder+fCMOS)
        EMCCD= imread(folder+fEMCCD)
    
        CMOS = CMOS[:,:CMOS.shape[1]//2]
    
        if not csv:
            # Produce csv - pass it to subroutine    
            pass
        else:
            fCMOScsv, fEMCCDcsv =  (filesCMOScsv[i],filesEMCCDcsv[i])
            # ReadROIS!? Show me csv files
            CMOS_pts = ReadROIs(loadtxt(fCMOS))
            EMCCD_pts  = ReadROIs(loadtxt(fEMCCD))
        
        #CenterROIs function
        CMOS_pts_center=CenterROIs(CMOS,CMOS_pts,ROIsize);
        EMCCD_pts_center=CenterROIs(EMCCD,EMCCD_pts,ROIsize);
        ptsC.append(CMOS_pts_center)
        ptsE.append(EMCCD_pts_center)

    ptsC = np.array(ptsC)
    ptsE = np.array(ptsE)
    Xc=ptsC[:,0]
    Yc=ptsC[:,1]
    Xe=ptsE[:,0]
    Ye=ptsE[:,1]

    fitobject = linregress(Xc,Xe)
    #slope, intercept, r_value, p_value, std_err
    ax, bx, _, _, _ = fitobject

    fitobject = linregress(Yc,Ye)
    #slope, intercept, r_value, p_value, std_err
    ax, bx, _, _, _ = fitobject

    return(ax,ay,bx,by)

File: matlab/function_EMCCD_CMOS_calib.m
function [ax ay bx by]=EMCCD_CMOS_calib(folder)
%folder='C:\Users\ludwig\Documents\MATLAB\18_11_29_pd23_11_div6_WIS_NR-BeRST\EMCCD-CMOS calib'; 
files = dir(fullfile(folder, '*.tif'));
files_csv = dir(fullfile(folder, '*.csv'));
%importing calib images
k=0;
n=1;
m=1;
for i=1:size(files) %real traces
    file = files(i).name;
    k=strfind(file, 'CMOS');
    if k~=0
        CMOS(:,:,n)=imread(fullfile(folder,file));
        n=n+1;
    else
        EMCCD(:,:,m)=imread(fullfile(folder,file));
        m=m+1;
    end 
end

CMOS=CMOS(:,1:size(CMOS,2)/2,:);
k=0;
n=1;
m=1;
for i=1:size(files_csv) %real traces
    file = files_csv(i).name;
    k=strfind(file, 'CMOS');
    if k~=0
        CMOS_pts{n}=ReadROIs(fullfile(folder,file));
        n=n+1;
    else
        EMCCD_pts{m}=ReadROIs(fullfile(folder,file));
        m=m+1;
    end 
end
 clearvars i file folder files k temp myvars k n m files_csv

% %Plotting images
% temp=squeeze(CMOS(1,:,:));
% img=imadjust(temp);
% figure;
% imshow(img)
% 
% clearvars img temp


% define ROIs
ROIsize=9;

for i=1:size(CMOS_pts,2)
    CMOS_pts_center{i}=CenterROIs(CMOS(:,:,i),CMOS_pts{i},ROIsize);
    EMCCD_pts_center{i}=CenterROIs(EMCCD(:,:,i),EMCCD_pts{i},ROIsize);
end
clearvars i CMOS_pts EMCCD_pts

%% Plot SumMoive with ROIs
% image=1;
% RB=0;
% norm=0;
% PlotROIs(CMOS(:,:,image),[CMOS_pts_center{image}],ROIsize,RB, norm)
% PlotROIs(EMCCD(:,:,image),[EMCCD_pts_center{image}],ROIsize,RB, norm)
% clearvars RB norm image

%%
ptsC=vertcat(CMOS_pts_center{:});
ptsE=vertcat(EMCCD_pts_center{:});
Xc=ptsC(:,1);
Yc=ptsC(:,2);
Xe=ptsE(:,1);
Ye=ptsE(:,2);

fitobject = fit(Xc,Xe,'poly1');
ax=fitobject.p1;
bx=fitobject.p2;

fitobject = fit(Yc,Ye,'poly1');
ay=fitobject.p1;
by=fitobject.p2;